bip-annex.mediawiki

```diff
@@ -0,0 +1,166 @@
+    {
+        CDataStream annex(MakeByteSpan(annex_stack), ANNEX_SER_TYPE, ANNEX_SER_VERSION);
+    
+        /* Pop annex tag */
+        uint8_t annex_tag;
+        annex.read(AsWritableBytes(Span{&annex_tag, 1}));
+        assert(annex_tag == ANNEX_TAG);
+    
+        /* Annex record are encoded as:
+         * `VarInt` + 1 bytes: type
+         * `VarInt` OR 0 bytes: length
+         * `length`-bytes: value
+         */
+        while (!annex_stack.empty()) {
+    
+            /* `VarInt`-bytes to encode the higher-mask type
+             * 1-byte to encode the lower-mask type
+             * 8192 records with 2 bytes.
+             * 1056767 records with 3 bytes.
+             * 40456331200 records with 4 bytes.
+             * See `VarIntMode` comment in serialize.h for more.
+             */
+            uint64_t nRecordHigherMask = 0;
+            annex >> VARINT(nRecordHigherMask);
+            uint8_t nRecordLowerMask;
+            annex.read(AsWritableBytes(Span{&nRecordLowerMask, 1}));
+    
+            const uint64_t nRecordType = nRecordHigherMask * 64 + (nRecordLowerMask & 0x3F);
+            size_t nRecordLength = 0;
+    
+            /* `VarInt`-bytes to encode the record length.
+             * 1-byte for 4 to 130 bytes of length.
+             * 2-byte for 131 to 16514 bytes of length.
+             * 3-byte for 16515 to 2113666 bytes of length.
+             * Special-case for record value of length 0, 1, 2, 3, bytes. Save 1 byte of record length.
+             */
+            if ((nRecordLowerMask & 0xc0) == 0) {
+                annex >> VARINT(nRecordLength);
+                nRecordLength += 3;
+            } else {
+                /* Special-case for record value of length 0, 1, 2, 3 bytes. Save 1 byte of record length */
+                nRecordLength = ((nRecordLowerMask & 0xc0) >> 6);
+            }
+    
+            /* Consensus rule : the annex is not short - end of data
+             * before reading is finished */
+            if (annex.size() < nRecordLength)
+                return false;
+    
+            std::vector<unsigned char> vRecordValue;
+            vRecordValue.resize(nRecordLength);
+            annex.read(AsWritableBytes(Span{&vRecordValue, nRecordLength}));
+    
+            switch (nRecordType) {
+                /* Consensus rule : record value must make sense, per
+                 * the tag spec */
+                default:
+                    return true;
+            }
+        }
+        return true;
+    }
+
+== Security ==
+
+=== DoSy annex ===
+
+Lengthy annex bytes stream could be given to nodes as a CPU DoS vector. Standard policy rules should be adequate
+to prevent that concern. If many annex fields are considered as valid and/or their validation is expensive, a
+compensation mechanism should be introduced to constrain witness producer to commit higher fees (e.g inflate witness
+weight in function of annex size).
```

Comment:
I think this is a little bit backwards -- I think it would be better to say that the annex should always be simple and fast to parse and verify (eg, only using information from the transaction, its utxos, and block headers; only requiring a single pass to parse) and that any expensive computation (such as signature validation) should be left for script evaluation.

Either way, this seems more like a "rationale" thing?